### ISSUE TITLE
RyuJIT could use mov reg32, reg32 to cast from uint to ulong

### ISSUE THREAD

**Issue by mikedn** [dotnet/coreclr#2714]
Casts from `uint` to `ulong` are implemented with `and reg32, 0xffffffff`:
```C#
class Program {
    static ulong ul;
    static void Main(string[] args) {
        ul = (ulong)(uint)args.Length + 42;
    }
}
```
generates
```
G_M44928_IG02:
       8B4108               mov      eax, dword ptr [rcx+8]
       25FFFFFFFF           and      eax, 0xFFFFFFFF
       4883C002             add      rax, 2
       488905DD49EAFF       mov      qword ptr [reloc classVar[0xffa25008]], rax
```
A shorter/faster `mov eax, eax` could be used instead.

Of course, no cast is actually needed because it was already done by the previous `mov`. But that's likely more complicated to detect and optimize.

**Comment by sejongoh** [comment]
Just out of curiosity, I looked into the JIT dump. MOV eax, .. clears the upper 32 bits of rax (FYI, MOV ax, ... doesn't seem to modify the upper bits of eax).

If we want to optimize this, what is a right approach in RyuJit? 1) Is it good for code generation to check how eax is produced and omit generating "AND eax, 0xffffffff"
? 2) Is a post-codegen optimization better? 3) any other way?

I am just curious about how instruction combining is handled in RyuJIT. In other compilers (not x64/x86 compilers), I experienced that doing codegen and optimizations together like 1) often caused unexpected and hard-to-imagine problems. So, I was wondering how RyuJIT generally handles optimizations like this.


```
Generating: N009 (  3,  3) [000002] ---X--------             *  indir     int    REG rax $c0
IN0001:        mov      eax, dword ptr [rcx+8]
Generating: N011 (  4,  5) [000004] ---X-----U--             *  cast      long <- ulong <- uint REG rax $100
IN0002:        and      eax, 0xFFFFFFFF
Generating: N013 (  1,  1) [000006] ------------             *  const     long   42 REG NA $140
Generating: N015 (  6,  7) [000007] ---X--------             *  +         long   REG rax $101
IN0003:        add      rax, 42
```

**Comment by mikedn** [comment]
> If we want to optimize this, what is a right approach in RyuJit? 1) Is it good for code generation to check how eax is produced and omit generating "AND eax, 0xffffffff"
 ? 2) Is a post-codegen optimization better? 3) any other way?

IMO if you want to completely get rid of such casts then the optimization would probably need to be done before register allocation. Otherwise you may end up with an unnecessary register allocation. Here's a slightly modified example:

```C#
uint x = (uint)args.Length;
ul = (ulong)x;
return (int)x;
```
which generates
```
G_M31374_IG02:
       8B4108               mov      eax, dword ptr [rcx+8]
       8BD0                 mov      edx, eax
       4889154C4AEAFF       mov      qword ptr [reloc classVar[0xa38b51f0]], rdx
G_M31374_IG03:
       C3                   ret
```
Two things can be observed:
1. There's no `and` instruction. Seems like it is generated only in certain cases.
2. `mov edx, eax` is pointless, the store that follows could have used `rax` directly.

It seems to me that if the optimization is done after register allocation then the `edx` register would likely still be seen as allocated even if the `mov` instruction is removed. So let's do it before regalloc. But how exactly can something like this be encoded in RyuJIT's IR? It's as if you have 2 LclVars of different types stored in the same register, hmm...

**Comment by JosephTremoulet** [comment]
> There's no and instruction. Seems like it is generated only in certain cases.

Yes, I'd expect the `and` you're seeing in the original example is from this code (in `CodeGen::getIntToIntCast` in codegenxarch.cpp):

```
        else if  (treeNode->gtSetFlags() && isUnsignedDst && castOp->InReg() && (targetReg == sourceReg))
        {
            // if we (might) need to set the flags and the value is in the same register
            // and we have an unsigned value then use AND instead of MOVZX
            noway_assert(ins == INS_movzx || ins == INS_mov);
            ins = INS_AND;
        }
```

which would explain why you're seeing it above in the example where the src/dst have the same register but not in the example where they don't.

**Comment by mikedn** [comment]
@JosephTremoulet Yes, I ran into that piece of code recently but I have no idea why would the flags be needed after the cast. The obvious case, comparing the result of the cast to 0, doesn't need this as the JIT figures out that it can discard the cast and compare the original integer value to 0. Wonder if isn't some fallout from `FEATURE_SET_FLAGS` which is not defined.

**Comment by JosephTremoulet** [comment]
@mikedn yeah, that was my take -- seems to me that `gtSetFlags()` is giving a conservative answer, and it would be good to refine that answer by taking the context into account to see if anyone will consume the flags, but communicating that information is what `FEATURE_SET_FLAGS` is for and that, as you mentioned, is not defined.

**Comment by mikedn** [comment]
It looks like codegenxarch's calls to `gtSetFlags` are completely useless. There's no `gtRequestSetFlags` call on cast nodes anywhere to be seen and anyway codegenxarch expects `GT_JTRUE`'s operand to be a compare operation, it asserts otherwise. `gtSetFlags`/`gtRequestSetFlags` are ultimately ARM specific and they shouldn't have been used in XArch specific code.

Enabling `FEATURE_SET_FLAGS` does fix this particular issue but removing `gtSetFlags` calls from codegenxarch seems more appropriate.
